test/test_cuda.py

```diff
@@ -4659,9 +4441,212 @@ def test_no_triton_on_import(self):
+
+            grads = [
+                [torch.randn_like(p) for p in params]
+                for _ in range(steps_warmup + steps_train)
+            ]
+
+            # Control (capturable=False)
+            if "capturable" in kwargs:
+                del kwargs["capturable"]
+            opt = optimizer_ctor(params_control, capturable=False, **kwargs)
+
+            for i in range(steps_warmup + steps_train):
+                for j, p in enumerate(params_control):
+                    p.grad = grads[i][j]
+                opt.step()
+
+            # capturable=True
+
+            opt = optimizer_ctor(params_graphed, capturable=True, **kwargs)
+
+            for i in range(steps_warmup):
+                for j, p in enumerate(params_graphed):
+                    p.grad = grads[i][j]
+                opt.step()
+
+            if actually_do_graphs:
+                g = torch.cuda.CUDAGraph()
+                with torch.cuda.graph(g):
+                    opt.step()
+
+            for i in range(steps_train):
+                if actually_do_graphs:
+                    for j, p in enumerate(params_graphed):
+                        p.grad.copy_(grads[i + steps_warmup][j])
+                    g.replay()
+                else:
+                    # Passing capturable=True to the constructor and running without graphs should still be
+                    # numerically correct, even if it's not ideal for performance.
+                    for j, p in enumerate(params_graphed):
+                        p.grad = grads[i + steps_warmup][j]
+                    opt.step()
+
+            for p_control, p_graphed in zip(params_control, params_graphed):
+                self.assertEqual(p_control, p_graphed)
+
+    @onlyCUDA
+    @unittest.skipIf(
+        not TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >=5.3 required for graphs"
+    )
+    @optims(
+        [
+            optim
+            for optim in optim_db
+            if optim.optim_cls
+            in [
+                torch.optim.NAdam,
+                torch.optim.RAdam,
+                torch.optim.Rprop,
+                torch.optim.Adam,
+                torch.optim.AdamW,
+                torch.optim.Adamax,
+                torch.optim.ASGD,
+                torch.optim.Adadelta,
+                torch.optim.RMSprop,
+            ]
+        ],
+        dtypes=[torch.float32],
+    )
+    def test_graph_optims(self, device, dtype, optim_info):
+        optim_cls = optim_info.optim_cls
+        all_optim_inputs = _get_optim_inputs_including_global_cliquey_kwargs(
+            device, dtype, optim_info, skip=("differentiable",)
+        )
+        has_betas = any(
+            "betas" in error_inp.optimizer_error_input.kwargs
+            for error_inp in optim_info.optim_error_inputs_func(
+                device="cpu", dtype=dtype
+            )
+        )
+        for optim_input in all_optim_inputs:
+            kwargs = optim_input.kwargs
+            if "lr" in kwargs:
```

Comment:
Why isn't a tensor lr okay for the tests?

Reply:
I think tensor lr works, but all the kwargs in the original test used lr=0.1 so I updated the lr here to be 0.1.

Reply:
Do you know if the bigger lr makes a difference or were you trying to faithfully replicate the configs? If the latter, it is confusing to have the kwarg specific change in the test case and it would be better to not include it unless there was a reason for having LR be bigger than normal.

Reply:
lr tensor for ```torch.optim.adam``` and ```torch.optim.adamw``` is not supported when capturable=False and foreach=True, and an error is thrown.